server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1943,6 +1951,7 @@ public RetentionLease addRetentionLease(
         Objects.requireNonNull(listener);
         assert assertPrimaryMode();
         verifyNotClosed();
+        ensureSoftDeletesEnabled("retention leases");
```

Comment:
I think you can add a test case for each of these cases (in IndexShardRetentionLeaseTests).

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/39922/commits/4b23b8b4452060c2d576875c487b6a2a2ed75078.